[카프카] 로또 2단계 미션 제출합니다.

include42 opened the pull request:
안녕하세요. 카프카입니다.

로또 2단계(수동 입력 구현) 다시 리뷰 요청 드립니다.

말씀해주신 내용들을 우선적으로 수정하였습니다.

특히 splitGame 로직에 대해 지적을 주셔서 고민을 많이 했는데요,
원래 기획은 [인스턴스를 둘로 나누는 메소드]였던지라
새로운 횟수 인스턴스를 만들고, 그 인스턴스가 가져간 값만큼 내 인스턴스의 필드 값을 깎는다,
라고 설계를 했었습니다.

다만 지적을 받고 나서 다시 생각해보니, 
이러면 굳이 횟수를 객체로 감싼 의미가 없겠다 싶었습니다.
그래서 LottoGameRepeat가 감싸고 있는 repeat 정수값을 final로 지정하고,
splitGame을 호출하면 횟수 인스턴스를 파라미터로 전달받아
그 값과 비교하여 나머지가 되는 값을 생성하여 반환하는 식으로 수정하였습니다.

이렇게 수정하니 로직도 좀 더 직관적이 되고,
컨트롤러 부분을 고치고 보니 기존에 난해했던 부분이 말끔해져 다행이라고 생각합니다.

혹시나 이보다 더 좋은 방법이 있거나, 혹은 이 방법이 좋지 못한 방법이라면
날카로운 지적 부탁드리겠습니다. 
고민하다가 오히려 돌아서 가는게 아닌지 걱정인 탓입니다.

이외에도 아직 부족한 부분이 많은 코드입니다.
잘못된 부분 말씀해주시면 고칠 수 있도록 노력하겠습니다.
감사합니다.

include42 committed:
[feat] 반복 횟수 저장을 위한 LottoGameRepeat 구현

LottoGameRepeat 게임 반복 횟수를 저장하고 반복 종료 여부를 검사하는 클래스
LottoGameRepeatTest 생성, 예외, 반복 종료 기능 테스트
Money 금액에 따른 로또 게임 횟수를 RepeatCount로 반환하도록 수정
ALL 구조 변경에 따른 코드 리팩토링

include42 committed:
[feat] LottoGameRepeat 내에 split로직 추가

LottoGameRepeat 수동/자동 게임 횟수 분할을 위한 split 로직 추가
LottoGameRepeatTest 예외사항 변경(0 미만만 예외로 인정)

include42 committed:
[feat] View 내에 수동 입력을 위한 메서드 추가

InputView 횟수 입력 및 수동 번호 입력 로직 추가
OutputView 횟수/수동 입력 포맷 출력 및 횟수 출력 메서드 오버라이딩

include42 committed:
[feat] 수동 입력 기능 구동하도록 구현

LottoController 수동 입력 메서드 생성 및 기능 작동 확인
LottoGame OutputView의 값 전달을 위한 Getter 메서드 생성
OutputView 로또 번호 출력 로직 변경(printLottoNumbers)
InputView 요구사항에 맞춰 일부 로직 변경

include42 committed:
[refactor] LottoController 구조 개선

LottoController 필드변수 없이 LottoGame을 makeLottoGame이 반환하도록 구현

include42 committed:
[refactor] LottoController 구조 개선

LottoController 수동/자동 중 하나가 0번 수행되어도 문제없도록 코드 수정
ALL 구조 변경에 따른 리팩토링

include42 committed:
[refactor] LottoGameRepeatTest에 대한 예외사항 테스트 추가

LottoGameRepeatTest splitLottoGameRepeat에 유효하지 않은 값이 전달될 시 예외 발생 테스트

include42 committed:
[refactor] LottoController 내의 랜덤생성 메소드를 LottoGame으로 이동

LottoGame LottoGenerator와 LottoGameRepeat를 받아 LottoTicket을 추가하는 메서드 구현
LottoController 수동 입력만 처리하고 자동입력은 LottoGame에 RandomNumberGenerator를 전달해서 수행하도록 변경
ALL 코드 컨벤션 준수를 위한 리팩토링

hongsii commented on src/main/java/domain/lottonumber/LottoGame.java:
```diff
@@ -11,17 +13,27 @@
         return lottoResult;
     }
+
+    public void makeLottoTickets(LottoGameRepeat autoGameRepeat, LottoGenerator randomLottoGenerator) {
```
인터페이스 구현체로 파라미터명을 지으셨네요. 적절한 이름으로 바꿔보면 좋을 것 같습니다.

---

수동으로 입력한 로또를 만들려면 사용하는 쪽에서 직접 티켓을 생성해 `add(..)` 를 호출하는 형태입니다.
반면에, 자동은 이 메서드를 사용해 생성할 수 있습니다. 이 클래스를 사용하는 방식을 동일하게 맞춰 보면 어떨까요?

include42 replied:
앗 컨트롤러에 있던 메서드를 옮기면서 이름이 그대로 들어갔었네요...!! 말도 안되는 실수를 했습니다ㅠㅠ
---
말씀해주신 수동/자동 부분은, 수동이 인풋뷰와 연결되어야 하는 부분 때문에 고치다 만 것처럼 되어버려서 고민하고 있었습니다. 일단 add 대신 makeLottoTickets를 오버라이딩하는 방향으로 수정해보겠습니다.

hongsii commented on src/main/java/domain/LottoGameRepeat.java:
```diff
@@ -0,0 +1,44 @@
+        validateSplitValue(repeat);
+        this.repeat -= repeat;
+        return new LottoGameRepeat(repeat);
+    }
```
메서드의 책임이 드러나도록 네이밍을 하신 것은 👍
코드를 그대로 옮긴 네이밍보다 도메인 객체의 성격이 드러나도록 도메인 요구사항과 관련된 네이밍으로 개선해보면 좋을 것 같습니다.

include42 replied:
원래는 split으로 지었다가 기능 파악이 어려워서, 지나치리 정직하게 이름을 지었습니다ㅎㅎ
게임의 종류가 바뀌어 나뉘는 것이니...splitGame정도면 괜찮을런지요? 일단 수정해서 반영해보겠습니다!!!

hongsii reviewed (changes requested):
2단계도 잘 구현해주셨습니다.
몇 가지 피드백드렸으니 확인하셔서 반영해주세요~

include42 committed:
[refactor] LottoGame의 add 메소드 삭제 및 로직변경

LottoGame add 없이 makeLottoTickets로 로직 통합
LottoController LottoGame 수정사항에 맞춰 로직 변경

include42 committed:
[refactor] LottoGameRepeat 내 메소드 이름 변경

LottoGameRepeat splitLottoGameRepeat->splitGame으로 변경

include42 committed:
[refactor] LottoRank 내의 calculateRank 로직 리팩토링

LottoRank calculateRank 내의 보너스 매칭 로직을 메서드 분리하여 가독성 강화

hongsii commented on src/main/java/domain/LottoGameRepeat.java:
```diff
@@ -0,0 +1,44 @@
+
+    public boolean checkLoopTerminate(int count) {
+        return count < repeat;
+    }
```
아래와 같은 네이밍으로 바꿔도 좋을 것 같네요.
자바에서는 관습적으로 boolean 타입의 네이밍을 할 때, is, has, can과 같은 접두사를 사용합니다.

`checkRepeatPositive` -> `hasRepeat`
`checkLoopTerminate ` -> `hasRepeat(int count)`

그리고  아래와 같이 중복을 제거할 수 있을 것 같네요

``` java
    public boolean checkRepeatPositive() {
        return checkLoopTerminate(0);
    }

    public boolean checkLoopTerminate(int count) {
        return count < repeat;
    }
```

hongsii commented on src/main/java/domain/LottoGameRepeat.java:
```diff
@@ -0,0 +1,44 @@
+        validateRepeatCount(repeat);
+        validateSplitValue(repeat);
+        this.repeat -= repeat;
+        return new LottoGameRepeat(repeat);
```
불변 객체로 만들려는 의도라고 생각되는데요. 이 객체는 현재 불변이 아닙니다.
그 이유는 새로운 객체를 만들기 전에 현재 객체의 상태를 바꾼 다음 그 값을 전달하고 있습니다.

hongsii commented on src/main/java/domain/lottonumber/LottoGame.java:
```diff
@@ -1,27 +1,39 @@
+        }
+    }
+
+    public void makeLottoTickets(LottoGenerator lottoGenerator) {
```
코멘트에 오버라이딩한다고 말씀해주셨는데 파라미터만 다른 같은 이름의 메서드를 만드는 것은 오버로딩이라고 합니다!
아마 헷갈리셨을거라 생각됩니다만, 중요한 개념이니 혹시나 해서 피드백드립니다!

include42 replied:
헉 오버라이딩이라고 했나요? 민망하네요ㅠㅠ
지적해주셔서 감사합니다!!

hongsii reviewed (changes requested):
피드백 반영 잘해주셨네요! 👍 
몇 가지 피드백드렸으니, 확인하셔서 반영해주세요~

include42 committed:
[refactor] LottoGameRepeat 내의 메소드 명 변경

LottoGameRepeat checkRepeatPositive,checkLoopTerminate를 hasRepeat로 변경 및 오버로딩

include42 committed:
[refactor] LottoGameRepeat splitGame 메소드 로직 수정

LottoGameRepeat 객체의 불변성을 위해, 필드변수 repeat을 final로 수정하고, splitGame이 값 수정 없이 새로운 인스턴스를 생성하게 하여 분리 로직을 더 자연스럽게 변경함
LottoController 파라미터로 전체 게임 횟수를 받아, 거기서 수동/자동을 나누도록 로직 직관화
ALL 코드 변경에 따른 리팩토링

hongsii reviewed (approved):
피드백 반영을 잘해주셨네요!
2단계까지 진행하시느라 고생하셨습니다!
머지하겠습니다 :)

include42 commented:
감사합니다!! 세심한 조언 주셔서 많이 배웠습니다ㅎㅎ
